README.md
# HISTORY

    5. 2013-04-17, add debug command in ubsutils.pipe.mr and MACROS in ubsutils.hce.decorator
    4. 2013-04-02, add ubsutils.pipe.router.Step
    3. 2013-03-17, add ubsutils.pipe.mr and ubsutils.pipe.router. add pyhce.py into ubstuls
    2. 2012-06-13, add get_rsv_list in parser.utils.
    1. 2012-03-02, beta release0.0.1: 完成将文本转化为内存词典 or shelve对象的工具函数。




# DESCRIPTION

    此python模块是提供ubs编程的一些小工具函数。设计目的是：
	
    1. 纯python
      1. 不包含 c/c++ 扩展，保证linux环境的服务器和windows的个人PC都可以使用。
    2. 仅包含应用广泛的小函数
      1. 与某些特定项目相关的工具函数不要放入此模块。


	下载： git clone git@github.com:jumpoutofworld/pt-utils.git



# INSTALL

    模块采用了setuptools管理安装。依赖setuptools, nose。
	
    1. python setup.py build
    2. python setup.py test            # 测试
    3. python setup.py -q bdist_egg    # 生成egg文件，供调用
    4. python setup.py install         # 安装入site-packages



# MORE INFO

	1. egg文件的使用
		1. egg是python的打包文件，类似java的jar包。适合上传供streaming程序调用。
		2. 使用方式 export PYTHONPATH=/path/2/egg/ubscake-0.0.2-py2.6.egg 或者将egg文件路径加入python的sys.path变量。
	2. api文档
       1. 项目采用epydoc自动生成api文档。
	   2. 在线版本见 http://bb-ps-ubs02.bb01.baidu.com:8080/ubs-api-doc/ubsutils/

# TODO

	1. 因为pyhce.cc和pyhce.py的设计不合理+不一致， pyhce的脚本必须from hecutils import emit和在文件最末尾使用 MAPPER_INPUT_FLAGSPLIT等，这不合理。推动pyhce.cc调整执行顺序，然后在mapper_setup中使用 MAPPER_INPUT_FLAGSPLIT, 恢复pyhce.py.
	2. 添加一些工具函数，
		1. 20131221，天和datetime对象相互转换。
		2. median函数
	3. filesplitter中的一系列函数，split_file_by_key，在处理极端数据（比如巨大的作弊cookie）内存会爆掉，提供参数控制每次返回的数量。-1不控制。 可以按照100， 1000等控制。
    4. pdbstreaming.py不支持readlines，非常annoying


HISTORY.md

# 2014-08-01, 星期五 14:18 

	1. 添加list tools：
	    def chunks(l, n):
		""" Yield successive n-sized chunks from l.
		"""
			for i in xrange(0, len(l), n):
				yield l[i:i + n]
	2. 添加flat2list


# 2014-01-27, 星期一 15:59

   1. 设计一套将table数据导入cube，一键分析的工具，参考方伟的实现。
   2. 参考argparse的ParseResult和urlparse的SplitResult，提供行对象的高效封装。
   3. 提供cube dense相关工具，
   4. 重写general_sum_reducer.py, 作为脚本工具



# 2013-08-06, 星期二 11:30

   1. hce模式处理命令行参数的函数
   2. debug模式获取单part作为输入，input path包含文件太多时也不好使，考虑改进
   3. 在命令行添加dfs参数
   

# HISTORY


   1. 2013-09-30, 星期一 12:48
       1. input路径中包含namenode信息，在debug抽取单part时，namenode信息会被丢掉，造成问题
           1. input = hdfs://szwg-rank-hdfs.dmop.baidu.com:54310/app/ps/rank/ubs/monitor-target-out/@date@/all/newquerycube     2. input = /app/ps/rank/ubs/monitor-target-out/@date@/all/newquerycube/part-00000
		   2. 解决： 利用urlsplit 解析url，将scheme和netloc添加在path前
       2. pdbstreaming.py input output script.py -c aa -d bbb
           1. pdbstreaming.py的参数解析过强，-c或者-d会报错
		   2. 解决： 第4个以后的参数截断，不由parse_args处理
       3. 如果job文件有严重错误
           1. debug step1 20130829 不报错： (mr):debug grep-weigou-in-holmes 20130829   ==== run test mrjob in 2013-08-30 15:12:  /home/work/pengtao/bin/ubs_mrframework/bin/Start.py ../conf/weigou.mod.conf ./tmp.default.holmes.job.conf.schedule.20130830-1511-28 -t 20130829 -p grep-weigou-in-holmes -s -test 
    	   2. run step1 -test 报错 [FATAL]@initEnv
		   3. 解决：debug内部调用 -test，没有打印-test返回的stdout和stderr。将这些信息打印到sys.stderr上。



ptutils/hce/step.py
#!/usr/bin/env python
#coding:utf-8
# Author:  pengtao --<pengtao@baidu.com>
# Purpose: 
#     1. the map/reduce job of hce framework
# History:
#     2. 2012/06/04 update
#         1. remove several high level step struct
#         2. add export_macros method for MRStep
#     1. 2012/02/03 created.

"""
可复用的pyhce框架的mapper/reducer代码。
"""


import sys
import datetime

try:
    # The pyhce.cc will do this for top level script automatically
    # Mannually do it here
    from hceutil import emit
except ImportError:
    pass


########################################################################
class MRStep:
    """abstract class for pyHCE map or reduce step.
    
    Three methods are list here. They are the counterparts of pyhce mapper/reducer. 
        1. step_setup
        2. step (required)
        3. step_cleanup
    Mapper/Reducer are not combined because they are totally isolated in MapReduce
    Framework.
    """

    #----------------------------------------------------------------------
    def __init__(self):
        """Constructor"""
        
    #----------------------------------------------------------------------
    def step_setup(self):
        """ """
        pass
    
    #----------------------------------------------------------------------
    def step(self, k, v):
        """ """
        raise NotImplementedError
    
    #----------------------------------------------------------------------
    def step_cleanup(self):
        """ """
        pass
    
    #----------------------------------------------------------------------
    def export_macros(self, local, type_):
        """
        export class method into global namespace

        @type local: dict
        @param local:  the local() object from parent scope
        @type type_: string
        @param type_:  "m"/"mapper" for mapper and "r"/"reducer" for reducer
        
        """
        
        if type_ in ('m', 'mapper'):
            local['mapper_setup'] = self.step_setup
            local['mapper'] = self.step
            local['mapper_cleanup'] = self.step_cleanup
        elif type_ in ('r', 'reducer'):
            local['reducer_setup'] = self.step_setup
            local['reducer'] = self.step
            local['reducer_cleanup'] = self.step_cleanup
        else:
            raise ValueError("unknown type: %s" % type_)
            
        return       
        
        
        
    
########################################################################
class MRStepCombinedParser(MRStep):
    """
    Load a parser, read the log line by line, _compute and put the result
    in _output.
    
    usage:
    =====
    
        Derive a new class from the Base
        
        >>> class Demo(MRStepCombineParser):
        >>>    def __init__(self):
        >>>        MRStepCombineParser.__init__(self, recordz.UIlogRecord, verbose=100)
        >>>
        >>>    def _compute(self):
        >>>        do some computing
        
        
    
    """

    #----------------------------------------------------------------------
    def __init__(self, cls, verbose_n=0):
        """
        @type cls: object
        @param cls: The record class object
        @type verbose_n: int
        @param verbose_n: print a mark to stderr every verbose_n records. Do not print if n==0.
        
        """
        self._parser_cls = cls
        self.verbose_n = verbose_n
        self._verbose_counter = 0
        
    #----------------------------------------------------------------------
    def step_setup(self):
        """
        """
        self.rec = self._parser_cls()
        self._output = {}
        
    #----------------------------------------------------------------------
    def _compute(self):
        """
        Compute a complete record.
          1. You can put the results in _output as combiner
          2. or emit directly.
          
          >>> self._output[k] = v
          >>> emit(k, v)
          
        """
        raise NotImplementedError
    
    #----------------------------------------------------------------------
    def step(self, k, v):
        """
        这里有一些问题，如果是多行的log，最后一个record没有计算，在step_cleanup中计算，单行日志会重复计算。
        """
        if self.verbose_n:
            self._verbose_counter += 1
            if self._verbose_counter % self.verbose_n == 0:
                print >> sys.stderr, "%d records done ..." % self._verbose_counter
            if self.rec.parseLine(v) == self.rec.LOG_FINISHED:
                self._compute()
        else:
            try:
                if self.rec.parseLine(v) == self.rec.LOG_FINISHED:
                    self._compute()
            except (KeyError, ValueError, IndexError):
                pass
            
     
    #----------------------------------------------------------------------
    def step_cleanup(self):
        """Compute last record and emit the data.
        """
        # last record for multi-line record. The processing is duplicated for single-line record.
        self._compute()
        
        for (k,v) in self._output.iteritems():
            if type(v) == type([]):
                emit(k, "\t".join(map(str,v)))
            else:
                emit(k, str(v))
                
        

        
def reducer_cat(k,vs):
    """
    simplely cat   
    """
    for v in vs:
        emit(k,str(v))
      

#----------------------------------------------------------------------
def reducer_simplesum(k, vs):
    """
    sum the counts of the key
    """
    s = 0
    for v in vs:
        s += float(v)
    emit(k, str(s))
    
#----------------------------------------------------------------------
def reducer_listsum(k, vs):
    """
    Sum the list values of the key.
    
    length of list is calculated from the first data line.
    
    Note that vs is of type hceutil.ReduceValues, which is not subscriptable.
    
    """
    s = []
    slen = 0
    for v in vs:
        delta = map(float, v.split("\t"))
        if s :
            for i in range(slen):
                s[i] += delta[i]
        else :
            s = delta
            slen = len(s)
    
    emit(k, "\t".join(map(str, s)))
    
#----------------------------------------------------------------------
def reducer_count(k, vs):
    """
    count the occurency of the key
    """
    c = 0
    for v in vs:
        c += 1
    emit(k, str(c))
    
